File: server/spring/src/main/java/com/ssafy/tadak/spring/common/exception/GlobalExceptionHandler.java
package com.ssafy.tadak.spring.common.exception;

import com.ssafy.tadak.spring.common.exception.dto.ErrorResponse;
import com.ssafy.tadak.spring.mattermost.infrastructure.MattermostNotifier;
import jakarta.servlet.http.HttpServletRequest;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.http.converter.HttpMessageNotReadableException;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;
import org.springframework.web.method.annotation.MethodArgumentTypeMismatchException;
import org.springframework.web.multipart.MaxUploadSizeExceededException;
import org.springframework.web.servlet.resource.NoResourceFoundException;

import java.util.Arrays;
import java.util.stream.Collectors;

@Slf4j
@RestControllerAdvice
@RequiredArgsConstructor
public class GlobalExceptionHandler {

    private static final String LOG_FORMAT = "request = {}, {} \n class = {} \n code = {} \n message = {}";

    @ExceptionHandler(GlobalException.class)
    public ResponseEntity<ErrorResponse> handleGlobalException(
            GlobalException exception,
            HttpServletRequest request
    ){
        ErrorCode errorCode = exception.getErrorCode();

        log.warn(
                LOG_FORMAT,
                request.getMethod(),
                request.getRequestURI(),
                exception.getClass().getSimpleName(),
                errorCode.code(),
                errorCode.message()
        );
        exception.printStackTrace();
        return ResponseEntity.status(exception.getHttpStatus()).body(ErrorResponse.from(errorCode));
    }

    @ExceptionHandler(MaxUploadSizeExceededException.class)
    public ResponseEntity<ErrorResponse> handleMaxSizeException(MaxUploadSizeExceededException e) {
        return ResponseEntity.status(413) // 413 Payload Too Large
                .body(new ErrorResponse("S4130", "업로드 가능한 파일 크기를 초과했습니다."));
    }

    @ExceptionHandler(MethodArgumentNotValidException.class)
    public ResponseEntity<ErrorResponse> handleBadRequestException(
            MethodArgumentNotValidException exception,
            HttpServletRequest request
    ) {
        String errorMessage = exception.getBindingResult()
                .getFieldErrors()
                .stream()
                .map(fieldError -> fieldError.getField() + ": " + fieldError.getDefaultMessage())
                .collect(Collectors.joining(", "));

        log.warn(
                LOG_FORMAT,
                request.getMethod(),
                request.getRequestURI(),
                exception.getClass().getSimpleName(),
                "BAD_REQUEST",
                errorMessage
        );
        exception.printStackTrace();

        return ResponseEntity
                .status(HttpStatus.BAD_REQUEST)
                .body(ErrorResponse.from(new ErrorCode("400", errorMessage)));
    }

    @ExceptionHandler({HttpMessageNotReadableException.class, MethodArgumentTypeMismatchException.class})
    public ResponseEntity<ErrorResponse> handleBadRequestExceptions(
            Exception exception,
            HttpServletRequest request
    ) {
        String errorMessage;
        if (exception instanceof MethodArgumentTypeMismatchException mismatchEx) {
            errorMessage = "요청 파라미터 타입이 잘못되었습니다: " + mismatchEx.getName();
        } else {
            errorMessage = "요청 본문을 읽을 수 없습니다. JSON 형식이나 필드 값을 확인하세요.";
        }

        log.warn(
                LOG_FORMAT,
                request.getMethod(),
                request.getRequestURI(),
                exception.getClass().getSimpleName(),
                "BAD_REQUEST",
                errorMessage
        );
        exception.printStackTrace();

        return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                .body(new ErrorResponse("400", errorMessage));
    }

    @ExceptionHandler(NoResourceFoundException.class)
    public ResponseEntity<ErrorResponse> handleNoResourceFound(
            NoResourceFoundException exception,
            HttpServletRequest request
    ) {
        log.warn(
                LOG_FORMAT,
                request.getMethod(),
                request.getRequestURI(),
                exception.getClass().getSimpleName(),
                HttpStatus.NOT_FOUND,
                "정적 리소스를 찾을 수 없습니다."
        );

        return ResponseEntity.status(HttpStatus.NOT_FOUND)
                .body(new ErrorResponse("404", "정적 리소스를 찾을 수 없습니다."));
    }

    @ExceptionHandler(Exception.class)
    public ResponseEntity<ErrorResponse> handleServerException(
            Exception exception,
            HttpServletRequest request
    ){
        log.warn(
                LOG_FORMAT,
                request.getMethod(),
                request.getRequestURI(),
                exception.getClass().getSimpleName(),
                HttpStatus.INTERNAL_SERVER_ERROR,
                exception.getMessage()
        );
        exception.printStackTrace();

        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                .body(new ErrorResponse("S5000", exception.getMessage()));
    }

    private String getStackTrace(Exception ex) {
        return Arrays.stream(ex.getStackTrace())
                .map(StackTraceElement::toString)
                .collect(Collectors.joining("\n"));
    }
}

File: server/auth_gateway/src/main/kotlin/exception/GlobalExceptionHandler.kt
package com.tadak.exception

import com.tadak.exception.GlobalException
import com.tadak.exception.dto.ErrorResponse
import io.ktor.server.application.*
import io.ktor.server.plugins.statuspages.*
import io.ktor.http.*
import io.ktor.server.plugins.*
import io.ktor.server.response.*
import io.ktor.server.request.*
import org.slf4j.LoggerFactory

fun Application.configureExceptionHandler() {

    val log = LoggerFactory.getLogger("GlobalExceptionHandler")

    install(StatusPages) {

        // GlobalException 핸들링
        exception<GlobalException> { call, cause ->
            val errorCode = cause.errorCode

            log.warn(
                """
                request = ${call.request.httpMethod.value} ${call.request.uri}
                class = ${cause::class.simpleName}
                code = ${errorCode.code}
                message = ${errorCode.message}
                """.trimIndent()
            )
            cause.printStackTrace()

            call.respond(
                cause.status,
                ErrorResponse.from(errorCode)
            )
        }

        // 요청 파싱 관련 예외 (ex: JSON 파싱 오류)
        exception<BadRequestException> { call, cause ->
            val errorMessage = cause.message ?: "Invalid request"

            log.warn(
                """
                request = ${call.request.httpMethod.value} ${call.request.uri}
                class = ${cause::class.simpleName}
                code = 400
                message = $errorMessage
                """.trimIndent()
            )
            cause.printStackTrace()

            call.respond(
                HttpStatusCode.BadRequest,
                ErrorResponse(code = "400", message = errorMessage)
            )
        }

        // 그 외 모든 Exception
        exception<Throwable> { call, cause ->
            log.warn(
                """
                request = ${call.request.httpMethod.value} ${call.request.uri}
                class = ${cause::class.simpleName}
                code = 500
                message = ${cause.message}
                """.trimIndent()
            )
            cause.printStackTrace()

            call.respond(HttpStatusCode.InternalServerError,
                ErrorResponse(code = "500", message = "서버에서 오류가 발생하였습니다. 잠시후 다시 시도해주세요"))
        }
    }
}